src/Nerdbank.MessagePack/Converters/ObjectArrayConverter`1.cs

```diff
@@ -412,193 +447,200 @@ int NextSyncBatchSize()
+						await reader.BufferNextStructuresAsync(syncBatchSize, syncBatchSize, context).ConfigureAwait(false);
+						MessagePackReader syncReader = reader.CreateBufferedReader();
+						for (int syncReadEndExclusive = i + syncBatchSize; i < syncReadEndExclusive; i++)
+						{
+							if (properties.Length > i && properties.Span[i]?.MsgPackReaders is var (deserialize, _))
+							{
+								deserialize(ref value, ref syncReader, context);
+							}
+							else if (unusedDataProperty?.Setter is not null)
+							{
+								unused ??= new();
+								unused.Add(i, syncReader.ReadRaw(context));
+							}
+							else
+							{
+								syncReader.Skip(context);
+							}
+						}
+
+						reader.ReturnReader(ref syncReader);
 					}
 
-					value = await deserializeAsync(value, reader, context).ConfigureAwait(false);
-				}
+					// Read any consecutive async properties.
+					for (; i < arrayLength && properties.Length > i; i++)
+					{
+						if (properties.Span[i] is not PropertyAccessors<T> { PreferAsyncSerialization: true, MsgPackReaders: (_, { } deserializeAsync) })
+						{
+							break;
+						}
 
-				int NextSyncReadBatchSize()
-				{
-					// We want to count the number of array elements need to be read up to the next async property.
-					for (int j = i; j < arrayLength; j++)
+						value = await deserializeAsync(value, reader, context).ConfigureAwait(false);
+					}
+
+					int NextSyncReadBatchSize()
 					{
-						if (properties.Length > j)
+						// We want to count the number of array elements need to be read up to the next async property.
+						for (int j = i; j < arrayLength; j++)
 						{
-							PropertyAccessors<T>? property = properties.Span[j];
-							if (property?.PreferAsyncSerialization is true && property.MsgPackReaders is not null)
+							if (properties.Length > j)
 							{
-								return j - i;
+								PropertyAccessors<T>? property = properties.Span[j];
+								if (property?.PreferAsyncSerialization is true && property.MsgPackReaders is not null)
+								{
+									return j - i;
+								}
 							}
 						}
-					}
 
-					// We didn't encounter any more async property readers.
-					return arrayLength - i;
+						// We didn't encounter any more async property readers.
+						return arrayLength - i;
+					}
 				}
 			}
-		}
 
-		if (unused is not null && value is not null && unusedDataProperty?.Setter is not null)
-		{
-			unusedDataProperty.Setter(ref value, unused);
-		}
+			if (unused is not null && value is not null && unusedDataProperty?.Setter is not null)
+			{
+				unusedDataProperty.Setter(ref value, unused);
+			}
 
-		callbacks?.OnAfterDeserialize();
+			callbacks?.OnAfterDeserialize();
 
-		return value;
+			return value;
+		}
+		catch (Exception ex) when (ShouldWrapSerializationException(ex, context.CancellationToken))
+		{
+			throw new MessagePackSerializationException($"An error occurred while deserializing object of type '{typeof(T).FullName}' from array format asynchronously.", ex);
```

Comment:
"from array format asynchronously" is too much detail. It's enough to say an error occurred while deserializing a given object type. Review all your changes and consolidate construction of these type-level strings to a `private protected` method on the shared base class.